dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/DefaultEventAnalyticsDimensionalItemService.java

```diff
@@ -0,0 +1,205 @@
+import static org.hisp.dhis.common.ValueType.IMAGE;
+import static org.hisp.dhis.common.ValueType.INTEGER;
+import static org.hisp.dhis.common.ValueType.INTEGER_NEGATIVE;
+import static org.hisp.dhis.common.ValueType.INTEGER_POSITIVE;
+import static org.hisp.dhis.common.ValueType.INTEGER_ZERO_OR_POSITIVE;
+import static org.hisp.dhis.common.ValueType.NUMBER;
+import static org.hisp.dhis.common.ValueType.PERCENTAGE;
+import static org.hisp.dhis.common.ValueType.TRACKER_ASSOCIATE;
+import static org.hisp.dhis.common.ValueType.TRUE_ONLY;
+import static org.hisp.dhis.common.ValueType.UNIT_INTERVAL;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+
+import org.hisp.dhis.analytics.event.EventAnalyticsDimensionalItemService;
+import org.hisp.dhis.analytics.event.EventsAnalyticsDimensionalItems;
+import org.hisp.dhis.category.Category;
+import org.hisp.dhis.category.CategoryCombo;
+import org.hisp.dhis.category.CategoryOptionGroupSet;
+import org.hisp.dhis.category.CategoryService;
+import org.hisp.dhis.common.BaseDimensionalItemObject;
+import org.hisp.dhis.common.DataDimensionType;
+import org.hisp.dhis.common.ValueType;
+import org.hisp.dhis.dataelement.DataElement;
+import org.hisp.dhis.program.Program;
+import org.hisp.dhis.program.ProgramStage;
+import org.hisp.dhis.program.ProgramStageService;
+import org.hisp.dhis.trackedentity.TrackedEntityAttribute;
+import org.springframework.stereotype.Service;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+@Service
+@RequiredArgsConstructor
+class DefaultEventAnalyticsDimensionalItemService implements EventAnalyticsDimensionalItemService
+{
+
+    final static Collection<ValueType> QUERY_DISALLOWED_VALUE_TYPES = ImmutableSet.of(
+        IMAGE,
+        FILE_RESOURCE,
+        TRACKER_ASSOCIATE );
+
+    final static Collection<ValueType> AGGREGATE_ALLOWED_VALUE_TYPES = ImmutableSet.of(
+        NUMBER,
+        UNIT_INTERVAL,
+        PERCENTAGE,
+        INTEGER,
+        INTEGER_POSITIVE,
+        INTEGER_NEGATIVE,
+        INTEGER_ZERO_OR_POSITIVE,
+        BOOLEAN,
+        TRUE_ONLY );
+
+    private final Map<OperationType, Predicate<ValueType>> OPERATION_FILTER = ImmutableMap.of(
+        OperationType.QUERY, not( QUERY_DISALLOWED_VALUE_TYPES::contains ),
+        OperationType.AGGREGATE, AGGREGATE_ALLOWED_VALUE_TYPES::contains );
+
+    @NonNull
+    private final ProgramStageService programStageService;
+
+    @NonNull
+    private final CategoryService categoryService;
+
+    @Override
+    public EventsAnalyticsDimensionalItems getQueryDimensionalItemsByProgramStageId( String programStageId )
+    {
+        return Optional.of( programStageId )
```

Comment:
Just wondering, how do handle sorting here? I.e. do we sort by "object type" and then name, or just by name?

Reply:
all dimensions are collected in a big heterogeneous collection, then this collection is sorted by specified field (in `PagingAndSortingService`, below) is any, or by `created` (as defined in `PagingAndSortingService`)